### ISSUE TITLE
windows_package failure in version_compare

### ISSUE THREAD

**Issue by maltewitt** [chef/chef#7681]
## Description

Given an already installed Windows Package **MyWindowsTool** at version **5.0.400.438**:
If you use **package** resource to upgrade this to version **5.0.400.439** this will fail with 

> Malformed version number string ["5.0.400.438"]

This is the currently installed version number. 

Looking at [/lib/chef/provider/package/windows.rb](https://github.com/chef/chef/blob/master/lib/chef/provider/package/windows.rb) (line 194) you find the following code:
```ruby
def version_compare(v1, v2)
          if v1 == "latest" || v2 == "latest"
            return 0
          end

          gem_v1 = Gem::Version.new(v1)
          gem_v2 = Gem::Version.new(v2)

          gem_v1 <=> gem_v2
end
```
The version number, which consists of four parts (major, minor, revision, build) is passed to the Gem::Version ctor. This assumes that versioning for windows applications matches the rules of Gem::Version, which is probably not always true, as in this example.

## Chef Version

Chef-Client 14.4.56

## Platform Version

Windows 10 (1803)

## Replication Case

```ruby
package "MyWindowsTool" do
    source "#{Chef::Config[:file_cache_path]}/VersionInfo.exe"
    version "5.0.400.439"
    installer_type :nsis
end
```

## Client Output

```
  * windows_package[MyWindowsTool] action install[2018-09-25T16:48:26+02:00] INFO: Processing windows_package[MyWindowsTool] action install (test_cookbook::test line 2)


    ================================================================================
    Error executing action `install` on resource 'windows_package[MyWindowsTool]'
    ================================================================================

    ArgumentError
    -------------
    Malformed version number string ["5.0.400.438"]

    Resource Declaration:
    ---------------------
    # In C:/chef/cache/cookbooks/test_cookbook/recipes/test.rb

      2: package "MyWindowsTool" do
      3:     source "#{Chef::Config[:file_cache_path]}/VersionInfo.exe"
      4:     version "5.0.400.439"
      5:     installer_type :nsis
      6: end

    Compiled Resource:
    ------------------
    # Declared in C:/chef/cache/cookbooks/test_cookbook/recipes/test.rb:2:in `from_file'

    windows_package("MyWindowsTool") do
      package_name "MyWindowsTool"
      action [:install]
      default_guard_interpreter :default
      declared_type :package
      cookbook_name "test_cookbook"
      recipe_name "test"
      source "c:\\chef\\cache\\versioninfo.exe"
      version "5.0.400.439"
      installer_type :nsis
    end

    System Info:
    ------------
    chef_version=14.4.56
    platform=windows
    platform_version=10.0.17134
    ruby=ruby 2.5.1p57 (2018-03-29 revision 63029) [x64-mingw32]
    program_name=C:/opscode/chef/bin/chef-client
    executable=C:/opscode/chef/bin/chef-client

[2018-09-25T16:48:26+02:00] INFO: Running queued delayed notifications before re-raising exception

Running handlers:
[2018-09-25T16:48:26+02:00] ERROR: Running exception handlers
Running handlers complete
[2018-09-25T16:48:26+02:00] ERROR: Exception handlers complete
Chef Client failed. 0 resources updated in 04 seconds
[2018-09-25T16:48:26+02:00] FATAL: Stacktrace dumped to C:/chef/cache/chef-stacktrace.out
[2018-09-25T16:48:26+02:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2018-09-25T16:48:26+02:00] FATAL: ArgumentError: windows_package[MyWindowsTool] (test_cookbook::test line 2) had an error: ArgumentError: Malformed version number string ["5.0.400.438"]
```

## Stacktrace

```
Generated at 2018-09-25 16:48:26 +0200
ArgumentError: windows_package[MyWindowsTool] (test_cookbook::test line 2) had an error: ArgumentError: Malformed version number string ["5.0.400.438"]
C:/opscode/chef/embedded/lib/ruby/2.5.0/rubygems/version.rb:208:in `initialize'
C:/opscode/chef/embedded/lib/ruby/2.5.0/rubygems/version.rb:199:in `new'
C:/opscode/chef/embedded/lib/ruby/2.5.0/rubygems/version.rb:199:in `new'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package/windows.rb:199:in `version_compare'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package.rb:507:in `block in target_version_array'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package.rb:590:in `block in each_package'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package.rb:586:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package.rb:586:in `each_with_index'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package.rb:586:in `each_package'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package.rb:478:in `target_version_array'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package.rb:86:in `block in <class:Package>'
(eval):2:in `block in action_install'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider.rb:236:in `instance_eval'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider.rb:236:in `compile_and_converge_action'
(eval):2:in `action_install'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package/windows.rb:139:in `action_install'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider.rb:182:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/resource.rb:577:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/runner.rb:70:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/runner.rb:98:in `block (2 levels) in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/runner.rb:98:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/runner.rb:98:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/runner.rb:97:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/client.rb:720:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/client.rb:715:in `catch'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/client.rb:715:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/client.rb:754:in `converge_and_save'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/client.rb:286:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/application.rb:303:in `run_with_graceful_exit_option'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/application.rb:279:in `block in run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/local_mode.rb:44:in `with_server_connectivity'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/application.rb:261:in `run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/application/client.rb:440:in `run_application'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/application.rb:66:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/bin/chef-client:25:in `<top (required)>'
C:/opscode/chef/bin/chef-client:75:in `load'
C:/opscode/chef/bin/chef-client:75:in `<main>'
```

**Comment by tas50** [comment]
Looks like we have another place where Gem won't cut it for version comparison. We had the same problem with Debian packages last year. Paging @stuartpreston @lamont-granquist for ideas.

**Comment by jordanmwright** [comment]
This issue is affecting me as well. It looks like this problem got introduced with commit https://github.com/chef/chef/commit/eba0385cdd993554c70caf8c25926b4e867f80fe as the `version_compare` method was not called in the `:install` action before that changeset. 
 
Trying to use `allow_downgrade true` as a workaround is not possible since the `windows_package` does not support the property.

**Comment by lamont-granquist** [comment]
- `version_compare` being called there is correct and necessary to address other bugs.
- windows_package should probably force `allow_downgrade` to always true in the resource?  does this even make any sense on windows?  i'd argue it makes very little sense in a declarative configuration management system to begin with and hence the issue i cut somewhere that we need to reverse the default here.
- @stuartpreston may be on the right track that the multipackage support in the superclass is being used incorrectly.

**Comment by jordanmwright** [comment]
I believe this issue is mis-categorized as `Sustaining Backlog`. This issue prevents upgrading Windows packages on all versions of Chef 14.4 and above. It is a blocking issue for me. I cannot upgrade past 14.3 until this is resolved.

**Comment by stuartpreston** [comment]
@jordanmwright Please could you help by adding your test case to this issue, what you’ve tried and what doesn’t work as expected? I

**Comment by jordanmwright** [comment]
@stuartpreston It is the same as the original report. When a Windows package is already installed and has a version in the registry (so `current_version` gets populated), the package will error on an upgrade.

In my test `1.2.3.4` is the installed version and `2.1.3.4` is the new version.  

Error: `Malformed version number string ["1.2.3.4"]`

Compiled Resource: 
```
windows_package("myapplication") do
  package_name "myapplication"
  action [:install]
  default_guard_interpreter :default
  declared_type :windows_package
  cookbook_name "mycookbook"
  recipe_name "myapprecipe"
  source "c:\\chef\\cache\\myapplication-2.1.3.4.exe"
  version "2.1.3.4"
  options "/S"
  installer_type :custom
  only_if { #code block }
end
```

Stacktrace:
```
C:/opscode/chef/embedded/lib/ruby/2.5.0/rubygems/version.rb:208:in `initialize'
C:/opscode/chef/embedded/lib/ruby/2.5.0/rubygems/version.rb:199:in `new'
C:/opscode/chef/embedded/lib/ruby/2.5.0/rubygems/version.rb:199:in `new'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package/windows.rb:199:in `version_compare'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package.rb:507:in `block in target_version_array'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package.rb:590:in `block in each_package'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package.rb:586:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package.rb:586:in `each_with_index'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package.rb:586:in `each_package'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package.rb:478:in `target_version_array'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package.rb:86:in `block in <class:Package>'
(eval):2:in `block in action_install'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider.rb:236:in `instance_eval'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider.rb:236:in `compile_and_converge_action'
(eval):2:in `action_install'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider/package/windows.rb:139:in `action_install'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/provider.rb:182:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/resource.rb:577:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/runner.rb:70:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/runner.rb:98:in `block (2 levels) in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/runner.rb:98:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/runner.rb:98:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/runner.rb:97:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/client.rb:720:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/client.rb:715:in `catch'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/client.rb:715:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/client.rb:754:in `converge_and_save'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/client.rb:286:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/application.rb:303:in `run_with_graceful_exit_option'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/application.rb:279:in `block in run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/local_mode.rb:44:in `with_server_connectivity'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/application.rb:261:in `run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/application/client.rb:440:in `run_application'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/lib/chef/application.rb:66:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.4.56-universal-mingw32/bin/chef-client:25:in `<top (required)>'
C:/opscode/chef/bin/chef-client:75:in `load'
C:/opscode/chef/bin/chef-client:75:in `<main>'
```

**Comment by stuartpreston** [comment]
@jordanmwright @maltewitt I have spent some time reviewing this and TL;DR Please can you try with Chef Client version 14.6.47 (latest stable) or higher?  I think we did have some issues with this in versions 14.4 and 14.5 but they should be resolved now and downgrades should be possible as long as the package allows it.

### How I repro'd

Attempted downgrade from version 7.5.9 to 7.5.8 via repro.rb:

```ruby
version_number = "7.5.8"
windows_package "Notepad++ (32-bit x86)" do
  source "c:/temp/npp.#{version_number}.Installer.exe"
  version version_number
end
```

This repro uses the public package Notepad++ as a source from:

1. https://notepad-plus-plus.org/repository/7.x/7.5.8/npp.7.5.8.Installer.exe
2. https://notepad-plus-plus.org/repository/7.x/7.5.9/npp.7.5.9.Installer.exe

## Outcome

### v14.4.56 - FAIL

```
C:\projects\chef\chef [master =]> git checkout v14.4.56
C:\projects\chef\chef [(v14.4.56)]> bundle exec chef-apply c:\temp\repro.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * windows_package[Notepad++ (32-bit x86)] action install

    ================================================================================
    Error executing action `install` on resource 'windows_package[Notepad++ (32-bit x86)]'
    ================================================================================

    ArgumentError
    -------------
    Malformed version number string ["7.5.9"]

    Resource Declaration:
    ---------------------
    # In c:\temp\repro.rb

      2: windows_package "Notepad++ (32-bit x86)" do
      3:   source "c:/temp/npp.#{version_number}.Installer.exe"
      4:   version version_number
      5: end

    Compiled Resource:
    ------------------
    # Declared in c:\temp\repro.rb:2:in `run_chef_recipe'

    windows_package("Notepad++ (32-bit x86)") do
      package_name "Notepad++ (32-bit x86)"
      action [:install]
      default_guard_interpreter :default
      declared_type :windows_package
      cookbook_name "(chef-apply cookbook)"
      recipe_name "(chef-apply recipe)"
      source "c:\\temp\\npp.7.5.8.installer.exe"
      version "7.5.8"
    end

    System Info:
    ------------
    chef_version=14.4.56
    platform=windows
    platform_version=10.0.17134
    ruby=ruby 2.5.1p57 (2018-03-29 revision 63029) [x64-mingw32]
    program_name=C:/Ruby25-x64/lib/ruby/gems/2.5.0/bin/chef-apply
    executable=C:/Ruby25-x64/lib/ruby/gems/2.5.0/bin/chef-apply

[2018-11-05T13:54:54+00:00] FATAL: Stacktrace dumped to C:/chef/cache/chef-stacktrace.out
[2018-11-05T13:54:54+00:00] FATAL: Stacktrace dumped to C:/chef/cache/chef-stacktrace.out
[2018-11-05T13:54:54+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2018-11-05T13:54:54+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2018-11-05T13:54:54+00:00] FATAL: ArgumentError: windows_package[Notepad++ (32-bit x86)] ((chef-apply cookbook)::(chef-apply recipe) line 2) had an error: ArgumentError: Malformed version number string ["7.5.9"]
[2018-11-05T13:54:54+00:00] FATAL: ArgumentError: windows_package[Notepad++ (32-bit x86)] ((chef-apply cookbook)::(chef-apply recipe) line 2) had an error: ArgumentError: Malformed version number string ["7.5.9"]
```

### 14.5.27 - FAIL

```
C:\projects\chef\chef [(v14.4.56)]> git checkout v14.5.27
Previous HEAD position was f93d85daa Bump version to 14.4.56 by Chef Expeditor
HEAD is now at ae5cb538c Bump version to 14.5.27 by Chef Expeditor
C:\projects\chef\chef [(v14.5.27)]> bundle exec chef-apply c:\temp\repro.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * windows_package[Notepad++ (32-bit x86)] action install

    ================================================================================
    Error executing action `install` on resource 'windows_package[Notepad++ (32-bit x86)]'
    ================================================================================

    ArgumentError
    -------------
    Malformed version number string ["7.5.9"]

    Resource Declaration:
    ---------------------
    # In c:\temp\repro.rb

      2: windows_package "Notepad++ (32-bit x86)" do
      3:   source "c:/temp/npp.#{version_number}.Installer.exe"
      4:   version version_number
      5: end

    Compiled Resource:
    ------------------
    # Declared in c:\temp\repro.rb:2:in `run_chef_recipe'

    windows_package("Notepad++ (32-bit x86)") do
      package_name "Notepad++ (32-bit x86)"
      action [:install]
      default_guard_interpreter :default
      declared_type :windows_package
      cookbook_name "(chef-apply cookbook)"
      recipe_name "(chef-apply recipe)"
      source "c:\\temp\\npp.7.5.8.installer.exe"
      version "7.5.8"
    end

    System Info:
    ------------
    chef_version=14.5.27
    platform=windows
    platform_version=10.0.17134
    ruby=ruby 2.5.1p57 (2018-03-29 revision 63029) [x64-mingw32]
    program_name=C:/Ruby25-x64/lib/ruby/gems/2.5.0/bin/chef-apply
    executable=C:/Ruby25-x64/lib/ruby/gems/2.5.0/bin/chef-apply

[2018-11-05T13:56:46+00:00] FATAL: Stacktrace dumped to C:/chef/cache/chef-stacktrace.out
[2018-11-05T13:56:46+00:00] FATAL: Stacktrace dumped to C:/chef/cache/chef-stacktrace.out
[2018-11-05T13:56:46+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2018-11-05T13:56:46+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2018-11-05T13:56:46+00:00] FATAL: ArgumentError: windows_package[Notepad++ (32-bit x86)] ((chef-apply cookbook)::(chef-apply recipe) line 2) had an error: ArgumentError: Malformed version number string ["7.5.9"]
[2018-11-05T13:56:46+00:00] FATAL: ArgumentError: windows_package[Notepad++ (32-bit x86)] ((chef-apply cookbook)::(chef-apply recipe) line 2) had an error: ArgumentError: Malformed version number string ["7.5.9"]
```

### 14.5.33 - FAIL

```
C:\projects\chef\chef [(v14.5.27)]> git checkout v14.5.33
Previous HEAD position was ae5cb538c Bump version to 14.5.27 by Chef Expeditor
HEAD is now at 30ae8c51d Bump version to 14.5.33 by Chef Expeditor
C:\projects\chef\chef [(v14.5.33)]> bundle exec chef-apply c:\temp\repro.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * windows_package[Notepad++ (32-bit x86)] action install

    ================================================================================
    Error executing action `install` on resource 'windows_package[Notepad++ (32-bit x86)]'
    ================================================================================

    ArgumentError
    -------------
    Malformed version number string ["7.5.9"]

    Resource Declaration:
    ---------------------
    # In c:\temp\repro.rb

      2: windows_package "Notepad++ (32-bit x86)" do
      3:   source "c:/temp/npp.#{version_number}.Installer.exe"
      4:   version version_number
      5: end

    Compiled Resource:
    ------------------
    # Declared in c:\temp\repro.rb:2:in `run_chef_recipe'

    windows_package("Notepad++ (32-bit x86)") do
      package_name "Notepad++ (32-bit x86)"
      action [:install]
      default_guard_interpreter :default
      declared_type :windows_package
      cookbook_name "(chef-apply cookbook)"
      recipe_name "(chef-apply recipe)"
      source "c:\\temp\\npp.7.5.8.installer.exe"
      version "7.5.8"
    end

    System Info:
    ------------
    chef_version=14.5.33
    platform=windows
    platform_version=10.0.17134
    ruby=ruby 2.5.1p57 (2018-03-29 revision 63029) [x64-mingw32]
    program_name=C:/Ruby25-x64/lib/ruby/gems/2.5.0/bin/chef-apply
    executable=C:/Ruby25-x64/lib/ruby/gems/2.5.0/bin/chef-apply

[2018-11-05T13:57:44+00:00] FATAL: Stacktrace dumped to C:/chef/cache/chef-stacktrace.out
[2018-11-05T13:57:44+00:00] FATAL: Stacktrace dumped to C:/chef/cache/chef-stacktrace.out
[2018-11-05T13:57:44+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2018-11-05T13:57:44+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2018-11-05T13:57:44+00:00] FATAL: ArgumentError: windows_package[Notepad++ (32-bit x86)] ((chef-apply cookbook)::(chef-apply recipe) line 2) had an error: ArgumentError: Malformed version number string ["7.5.9"]
[2018-11-05T13:57:44+00:00] FATAL: ArgumentError: windows_package[Notepad++ (32-bit x86)] ((chef-apply cookbook)::(chef-apply recipe) line 2) had an error: ArgumentError: Malformed version number string ["7.5.9"]
```

### 14.6.47 - PASS

```
C:\projects\chef\chef [(v14.5.33)]> git checkout v14.6.47
Previous HEAD position was 30ae8c51d Bump version to 14.5.33 by Chef Expeditor
HEAD is now at 961596c87 Bump version to 14.6.47 by Chef Expeditor
C:\projects\chef\chef [(v14.6.47)]> bundle exec chef-apply c:\temp\repro.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * windows_package[Notepad++ (32-bit x86)] action install
    - install version 7.5.8 of package Notepad++ (32-bit x86)
```

**Comment by jordanmwright** [comment]
@stuartpreston I believe the issue is bypassed in 14.6.47 but not fixed. 

Looking through the change log of 14.6.47, the default `allow_downgrade` was changed from `false` to `true` in https://github.com/chef/chef/pull/7701. Since `windows_package` has no property of `allow_downgrade` the value will always be `true`. The if statement in question allows short circuiting so the `version_compare` function is never called. 

`elsif current_version && !allow_downgrade && version_compare(current_version, new_version) == 1`

So the error is not being thrown in 14.6.47, but the bug does still exist.

**Comment by stuartpreston** [comment]
Ok I’m getting a bit confused here, my testing showed 14.6.47 to be idempotent (marked install of the same version as up to date) and supported upgrade and downgrade to the specified version. 

Irrespective of your understanding of the code could you comment on the bug from a user perspective?

**Comment by jordanmwright** [comment]
I tested 14.7.17 and it works as expected. I tested upgrades, downgrades and no version change.
